File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/20．联合战斗成员生命周期.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__Delete = ____lualib.__TS__Delete
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local jass = require("jass.common")
local GetHandleId = jass.GetHandleId
local function _____590D_5236_72B6_6001_5217_8868(_____5217_8868)
    local result = {}
    do
        local i = 0
        while i < #_____5217_8868 do
            result[#result + 1] = _____5217_8868[i + 1]
            i = i + 1
        end
    end
    return result
end
local function _____72B6_6001_5728_5217_8868_4E2D(_____72B6_6001, _____5217_8868)
    do
        local i = 0
        while i < #_____5217_8868 do
            if _____5217_8868[i + 1] == _____72B6_6001 then
                return true
            end
            i = i + 1
        end
    end
    return false
end
local function _____53D6_5355_4F4D_53E5_67C4ID(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return 0
    end
    return GetHandleId(_____5355_4F4D) or 0
end
local _____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0 = __TS__Class()
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.name = "联合战斗成员生命周期实现"
function _____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype.____constructor(self, _____53C2_6570)
    self["成员表"] = {}
    self["成员Key列表"] = {}
    self["单位到成员Key"] = {}
    self["最终结算已触发"] = false
    self["已销毁"] = false
    self["参数"] = _____53C2_6570
    self["名称"] = _____53C2_6570["名称"]
    self["默认最终状态列表"] = _____590D_5236_72B6_6001_5217_8868(_____53C2_6570["默认最终状态列表"] or ({"崩解", "倒地", "离场"}))
    local _____6210_5458_5217_8868 = _____53C2_6570["成员列表"] or ({})
    do
        local i = 0
        while i < #_____6210_5458_5217_8868 do
            self["登记成员"](self, _____6210_5458_5217_8868[i + 1])
            i = i + 1
        end
    end
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["登记成员"] = function(self, _____5B9A_4E49)
    if self["已销毁"] or _____5B9A_4E49.key == "" or self["成员表"][_____5B9A_4E49.key] ~= nil then
        return false
    end
    local _____6210_5458 = {
        key = _____5B9A_4E49.key,
        ["单位"] = _____5B9A_4E49["单位"],
        ["角色"] = _____5B9A_4E49["角色"],
        ["状态"] = _____5B9A_4E49["初始状态"] or "未加入",
        ["数据"] = _____5B9A_4E49["数据"],
        ["参与最终结算"] = _____5B9A_4E49["参与最终结算"],
        ["最终状态列表"] = _____590D_5236_72B6_6001_5217_8868(_____5B9A_4E49["最终状态列表"] or self["默认最终状态列表"])
    }
    self["成员表"][_____5B9A_4E49.key] = _____6210_5458
    local ____self__6210_5458Key_5217_8868_0 = self["成员Key列表"]
    ____self__6210_5458Key_5217_8868_0[#____self__6210_5458Key_5217_8868_0 + 1] = _____5B9A_4E49.key
    self["登记单位映射"](self, _____6210_5458)
    return true
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["移除成员"] = function(self, key)
    if self["已销毁"] then
        return false
    end
    local _____6210_5458 = self["成员表"][key]
    if _____6210_5458 == nil then
        return false
    end
    self["移除单位映射"](self, _____6210_5458)
    __TS__Delete(self["成员表"], key)
    do
        local i = 0
        while i < #self["成员Key列表"] do
            do
                if self["成员Key列表"][i + 1] ~= key then
                    goto __continue20
                end
                __TS__ArraySplice(self["成员Key列表"], i, 1)
                break
            end
            ::__continue20::
            i = i + 1
        end
    end
    return true
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["更新单位"] = function(self, key, _____5355_4F4D)
    if self["已销毁"] then
        return false
    end
    local _____6210_5458 = self["成员表"][key]
    if _____6210_5458 == nil then
        return false
    end
    self["移除单位映射"](self, _____6210_5458)
    _____6210_5458["单位"] = _____5355_4F4D
    self["登记单位映射"](self, _____6210_5458)
    return true
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["设置状态"] = function(self, key, _____72B6_6001, _____539F_56E0)
    if _____539F_56E0 == nil then
        _____539F_56E0 = "成员状态变化"
    end
    if self["已销毁"] then
        return false
    end
    local _____6210_5458 = self["成员表"][key]
    if _____6210_5458 == nil or _____72B6_6001 == "" or _____6210_5458["状态"] == _____72B6_6001 then
        return false
    end
    local _____65E7_72B6_6001 = _____6210_5458["状态"]
    _____6210_5458["状态"] = _____72B6_6001
    if self["参数"]["on状态变化"] ~= nil then
        self["参数"]["on状态变化"]({["成员"] = _____6210_5458, ["旧状态"] = _____65E7_72B6_6001, ["新状态"] = _____72B6_6001, ["原因"] = _____539F_56E0})
    end
    self["检查最终结算"](self)
    return true
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["写入数据"] = function(self, key, _____6570_636E)
    if self["已销毁"] then
        return false
    end
    local _____6210_5458 = self["成员表"][key]
    if _____6210_5458 == nil then
        return false
    end
    _____6210_5458["数据"] = _____6570_636E
    return true
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["取成员"] = function(self, key)
    return self["成员表"][key]
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["按单位取成员"] = function(self, _____5355_4F4D)
    local id = _____53D6_5355_4F4D_53E5_67C4ID(_____5355_4F4D)
    if id == 0 then
        return nil
    end
    local key = self["单位到成员Key"][id]
    local ____temp_1
    if key == nil then
        ____temp_1 = nil
    else
        ____temp_1 = self["成员表"][key]
    end
    return ____temp_1
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["取成员列表"] = function(self)
    local result = {}
    do
        local i = 0
        while i < #self["成员Key列表"] do
            local _____6210_5458 = self["成员表"][self["成员Key列表"][i + 1]]
            if _____6210_5458 ~= nil then
                result[#result + 1] = _____6210_5458
            end
            i = i + 1
        end
    end
    return result
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["按角色取成员"] = function(self, _____89D2_8272)
    local result = {}
    local _____6210_5458_5217_8868 = self["取成员列表"](self)
    do
        local i = 0
        while i < #_____6210_5458_5217_8868 do
            if _____6210_5458_5217_8868[i + 1]["角色"] == _____89D2_8272 then
                result[#result + 1] = _____6210_5458_5217_8868[i + 1]
            end
            i = i + 1
        end
    end
    return result
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["按状态取成员"] = function(self, _____72B6_6001)
    local result = {}
    local _____6210_5458_5217_8868 = self["取成员列表"](self)
    do
        local i = 0
        while i < #_____6210_5458_5217_8868 do
            if _____6210_5458_5217_8868[i + 1]["状态"] == _____72B6_6001 then
                result[#result + 1] = _____6210_5458_5217_8868[i + 1]
            end
            i = i + 1
        end
    end
    return result
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["任一满足"] = function(self, _____5224_65AD)
    local _____6210_5458_5217_8868 = self["取成员列表"](self)
    do
        local i = 0
        while i < #_____6210_5458_5217_8868 do
            if _____5224_65AD(_____6210_5458_5217_8868[i + 1]) then
                return true
            end
            i = i + 1
        end
    end
    return false
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["全部满足"] = function(self, _____5224_65AD, _____53EA_770B_6700_7EC8_7ED3_7B97_6210_5458)
    if _____53EA_770B_6700_7EC8_7ED3_7B97_6210_5458 == nil then
        _____53EA_770B_6700_7EC8_7ED3_7B97_6210_5458 = false
    end
    local _____6210_5458_5217_8868 = self["取成员列表"](self)
    local _____5DF2_68C0_67E5_6570_91CF = 0
    do
        local i = 0
        while i < #_____6210_5458_5217_8868 do
            do
                local _____6210_5458 = _____6210_5458_5217_8868[i + 1]
                if _____53EA_770B_6700_7EC8_7ED3_7B97_6210_5458 and not _____6210_5458["参与最终结算"] then
                    goto __continue53
                end
                _____5DF2_68C0_67E5_6570_91CF = _____5DF2_68C0_67E5_6570_91CF + 1
                if not _____5224_65AD(_____6210_5458) then
                    return false
                end
            end
            ::__continue53::
            i = i + 1
        end
    end
    return _____5DF2_68C0_67E5_6570_91CF > 0
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["是否满足最终结算"] = function(self)
    return self["全部满足"](
        self,
        function(_____6210_5458)
            return _____72B6_6001_5728_5217_8868_4E2D(_____6210_5458["状态"], _____6210_5458["最终状态列表"])
        end,
        true
    )
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["已触发最终结算"] = function(self)
    return self["最终结算已触发"]
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["销毁"] = function(self)
    if self["已销毁"] then
        return
    end
    self["已销毁"] = true
    self["成员表"] = {}
    self["成员Key列表"] = {}
    self["单位到成员Key"] = {}
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["检查最终结算"] = function(self)
    if self["最终结算已触发"] or not self["是否满足最终结算"](self) then
        return
    end
    self["最终结算已触发"] = true
    if self["参数"]["on满足最终结算"] ~= nil then
        self["参数"]["on满足最终结算"](self["取成员列表"](self))
    end
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["登记单位映射"] = function(self, _____6210_5458)
    local id = _____53D6_5355_4F4D_53E5_67C4ID(_____6210_5458["单位"])
    if id ~= 0 then
        self["单位到成员Key"][id] = _____6210_5458.key
    end
end
_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0.prototype["移除单位映射"] = function(self, _____6210_5458)
    local id = _____53D6_5355_4F4D_53E5_67C4ID(_____6210_5458["单位"])
    if id ~= 0 and self["单位到成员Key"][id] == _____6210_5458.key then
        __TS__Delete(self["单位到成员Key"], id)
    end
end
____exports["创建联合战斗成员生命周期"] = function(_____53C2_6570)
    local _____5B9E_4F8B = __TS__New(_____8054_5408_6218_6597_6210_5458_751F_547D_5468_671F_5B9E_73B0, _____53C2_6570)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_2 = _____53C2_6570["清理"]
        ____self_2["登记清理"](
            ____self_2,
            _____53C2_6570["名称"] .. "-联合成员生命周期",
            function()
                _____5B9E_4F8B["销毁"](_____5B9E_4F8B)
            end
        )
    end
    return _____5B9E_4F8B
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/20．联合战斗成员生命周期.ts
/** @noSelfInFile */

import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";

const jass = require("jass.common") as any;
const GetHandleId = jass.GetHandleId as (handle: any) => number;

export type 联合战斗成员角色 = "主目标" | "护卫" | "搭档" | "机制单位" | (string & {});
export type 联合战斗成员状态 = "未加入" | "活跃" | "失衡" | "锁血" | "崩解" | "倒地" | "离场" | (string & {});

export interface 联合战斗成员定义<TData = Record<string, any>> {
  key: string;
  单位?: any;
  角色: 联合战斗成员角色;
  初始状态?: 联合战斗成员状态;
  数据?: TData;
  参与最终结算: boolean;
  最终状态列表?: 联合战斗成员状态[];
}

export interface 联合战斗成员<TData = Record<string, any>> {
  key: string;
  单位?: any;
  角色: 联合战斗成员角色;
  状态: 联合战斗成员状态;
  数据?: TData;
  参与最终结算: boolean;
  最终状态列表: 联合战斗成员状态[];
}

export interface 联合战斗成员状态变化事件<TData = Record<string, any>> {
  成员: 联合战斗成员<TData>;
  旧状态: 联合战斗成员状态;
  新状态: 联合战斗成员状态;
  原因: string;
}

export interface 联合战斗成员生命周期参数<TData = Record<string, any>> {
  名称: string;
  清理?: 机制清理篮子;
  成员列表?: 联合战斗成员定义<TData>[];
  默认最终状态列表?: 联合战斗成员状态[];
  on状态变化?: (this: void, event: 联合战斗成员状态变化事件<TData>) => void;
  on满足最终结算?: (this: void, 成员列表: 联合战斗成员<TData>[]) => void;
}

export interface 联合战斗成员生命周期<TData = Record<string, any>> {
  readonly 名称: string;
  登记成员(定义: 联合战斗成员定义<TData>): boolean;
  移除成员(key: string): boolean;
  更新单位(key: string, 单位?: any): boolean;
  设置状态(key: string, 状态: 联合战斗成员状态, 原因?: string): boolean;
  写入数据(key: string, 数据?: TData): boolean;
  取成员(key: string): 联合战斗成员<TData> | undefined;
  按单位取成员(单位: any): 联合战斗成员<TData> | undefined;
  取成员列表(): 联合战斗成员<TData>[];
  按角色取成员(角色: 联合战斗成员角色): 联合战斗成员<TData>[];
  按状态取成员(状态: 联合战斗成员状态): 联合战斗成员<TData>[];
  任一满足(判断: (this: void, 成员: 联合战斗成员<TData>) => boolean): boolean;
  全部满足(判断: (this: void, 成员: 联合战斗成员<TData>) => boolean, 只看最终结算成员?: boolean): boolean;
  是否满足最终结算(): boolean;
  已触发最终结算(): boolean;
  销毁(): void;
}

function 复制状态列表(this: void, 列表: 联合战斗成员状态[]): 联合战斗成员状态[] {
  const result: 联合战斗成员状态[] = [];
  for (let i = 0; i < 列表.length; i++) result.push(列表[i]);
  return result;
}

function 状态在列表中(this: void, 状态: 联合战斗成员状态, 列表: 联合战斗成员状态[]): boolean {
  for (let i = 0; i < 列表.length; i++) {
    if (列表[i] === 状态) return true;
  }
  return false;
}

function 取单位句柄ID(this: void, 单位: any): number {
  if (单位 == null || 单位 === 0) return 0;
  return GetHandleId(单位) || 0;
}

class 联合战斗成员生命周期实现<TData> implements 联合战斗成员生命周期<TData> {
  readonly 名称: string;
  private 参数: 联合战斗成员生命周期参数<TData>;
  private 默认最终状态列表: 联合战斗成员状态[];
  private 成员表: Record<string, 联合战斗成员<TData> | undefined> = {};
  private 成员Key列表: string[] = [];
  private 单位到成员Key: Record<number, string | undefined> = {};
  private 最终结算已触发 = false;
  private 已销毁 = false;

  constructor(参数: 联合战斗成员生命周期参数<TData>) {
    this.参数 = 参数;
    this.名称 = 参数.名称;
    this.默认最终状态列表 = 复制状态列表(参数.默认最终状态列表 ?? ["崩解", "倒地", "离场"]);
    const 成员列表 = 参数.成员列表 ?? [];
    for (let i = 0; i < 成员列表.length; i++) this.登记成员(成员列表[i]);
  }

  登记成员(定义: 联合战斗成员定义<TData>): boolean {
    if (this.已销毁 || 定义.key === "" || this.成员表[定义.key] != null) return false;
    const 成员: 联合战斗成员<TData> = {
      key: 定义.key,
      单位: 定义.单位,
      角色: 定义.角色,
      状态: 定义.初始状态 ?? "未加入",
      数据: 定义.数据,
      参与最终结算: 定义.参与最终结算,
      最终状态列表: 复制状态列表(定义.最终状态列表 ?? this.默认最终状态列表),
    };
    this.成员表[定义.key] = 成员;
    this.成员Key列表.push(定义.key);
    this.登记单位映射(成员);
    return true;
  }

  移除成员(key: string): boolean {
    if (this.已销毁) return false;
    const 成员 = this.成员表[key];
    if (成员 == null) return false;
    this.移除单位映射(成员);
    delete this.成员表[key];
    for (let i = 0; i < this.成员Key列表.length; i++) {
      if (this.成员Key列表[i] !== key) continue;
      this.成员Key列表.splice(i, 1);
      break;
    }
    return true;
  }

  更新单位(key: string, 单位?: any): boolean {
    if (this.已销毁) return false;
    const 成员 = this.成员表[key];
    if (成员 == null) return false;
    this.移除单位映射(成员);
    成员.单位 = 单位;
    this.登记单位映射(成员);
    return true;
  }

  设置状态(key: string, 状态: 联合战斗成员状态, 原因: string = "成员状态变化"): boolean {
    if (this.已销毁) return false;
    const 成员 = this.成员表[key];
    if (成员 == null || 状态 === "" || 成员.状态 === 状态) return false;
    const 旧状态 = 成员.状态;
    成员.状态 = 状态;
    if (this.参数.on状态变化 != null) this.参数.on状态变化({ 成员, 旧状态, 新状态: 状态, 原因 });
    this.检查最终结算();
    return true;
  }

  写入数据(key: string, 数据?: TData): boolean {
    if (this.已销毁) return false;
    const 成员 = this.成员表[key];
    if (成员 == null) return false;
    成员.数据 = 数据;
    return true;
  }

  取成员(key: string): 联合战斗成员<TData> | undefined {
    return this.成员表[key];
  }

  按单位取成员(单位: any): 联合战斗成员<TData> | undefined {
    const id = 取单位句柄ID(单位);
    if (id === 0) return undefined;
    const key = this.单位到成员Key[id];
    return key == null ? undefined : this.成员表[key];
  }

  取成员列表(): 联合战斗成员<TData>[] {
    const result: 联合战斗成员<TData>[] = [];
    for (let i = 0; i < this.成员Key列表.length; i++) {
      const 成员 = this.成员表[this.成员Key列表[i]];
      if (成员 != null) result.push(成员);
    }
    return result;
  }

  按角色取成员(角色: 联合战斗成员角色): 联合战斗成员<TData>[] {
    const result: 联合战斗成员<TData>[] = [];
    const 成员列表 = this.取成员列表();
    for (let i = 0; i < 成员列表.length; i++) {
      if (成员列表[i].角色 === 角色) result.push(成员列表[i]);
    }
    return result;
  }

  按状态取成员(状态: 联合战斗成员状态): 联合战斗成员<TData>[] {
    const result: 联合战斗成员<TData>[] = [];
    const 成员列表 = this.取成员列表();
    for (let i = 0; i < 成员列表.length; i++) {
      if (成员列表[i].状态 === 状态) result.push(成员列表[i]);
    }
    return result;
  }

  任一满足(判断: (this: void, 成员: 联合战斗成员<TData>) => boolean): boolean {
    const 成员列表 = this.取成员列表();
    for (let i = 0; i < 成员列表.length; i++) {
      if (判断(成员列表[i])) return true;
    }
    return false;
  }

  全部满足(
    判断: (this: void, 成员: 联合战斗成员<TData>) => boolean,
    只看最终结算成员: boolean = false,
  ): boolean {
    const 成员列表 = this.取成员列表();
    let 已检查数量 = 0;
    for (let i = 0; i < 成员列表.length; i++) {
      const 成员 = 成员列表[i];
      if (只看最终结算成员 && !成员.参与最终结算) continue;
      已检查数量 += 1;
      if (!判断(成员)) return false;
    }
    return 已检查数量 > 0;
  }

  是否满足最终结算(): boolean {
    return this.全部满足(function 联合战斗最终状态判断(this: void, 成员: 联合战斗成员<TData>): boolean {
      return 状态在列表中(成员.状态, 成员.最终状态列表);
    }, true);
  }

  已触发最终结算(): boolean {
    return this.最终结算已触发;
  }

  销毁(): void {
    if (this.已销毁) return;
    this.已销毁 = true;
    this.成员表 = {};
    this.成员Key列表 = [];
    this.单位到成员Key = {};
  }

  private 检查最终结算(): void {
    if (this.最终结算已触发 || !this.是否满足最终结算()) return;
    this.最终结算已触发 = true;
    if (this.参数.on满足最终结算 != null) this.参数.on满足最终结算(this.取成员列表());
  }

  private 登记单位映射(成员: 联合战斗成员<TData>): void {
    const id = 取单位句柄ID(成员.单位);
    if (id !== 0) this.单位到成员Key[id] = 成员.key;
  }

  private 移除单位映射(成员: 联合战斗成员<TData>): void {
    const id = 取单位句柄ID(成员.单位);
    if (id !== 0 && this.单位到成员Key[id] === 成员.key) delete this.单位到成员Key[id];
  }
}

export function 创建联合战斗成员生命周期<TData = Record<string, any>>(
  this: void,
  参数: 联合战斗成员生命周期参数<TData>,
): 联合战斗成员生命周期<TData> {
  const 实例 = new 联合战斗成员生命周期实现(参数);
  if (参数.清理 != null) {
    参数.清理.登记清理(参数.名称 + "-联合成员生命周期", function 联合战斗成员生命周期清理(this: void): void {
      实例.销毁();
    });
  }
  return 实例;
}
